src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs

```diff
@@ -13,8 +9,49 @@ internal partial class LocalRewriter
     {
         public override BoundNode VisitTupleCreationExpression(BoundTupleCreationExpression node)
         {
-            var rewrittenArguments = VisitList(node.Arguments);
-            return new BoundObjectCreationExpression(node.Syntax, node.ConstructorOpt, rewrittenArguments);
+            ImmutableArray<BoundExpression> rewrittenArguments = VisitList(node.Arguments);
+            return RewriteTupleCreationExpression(node, rewrittenArguments);
+        }
+
+        /// <summary>
+        /// Converts the expression for creating a tuple instance into an expression creating a ValueTuple (if short) or nested ValueTuples (if longer).
+        ///
+        /// For instance, for a long tuple we'll generate:
+        /// creationExpression(ctor=largestCtor, args=firstArgs+(nested creationExpression for remainder, with smaller ctor and next few args))
+        /// </summary>
+        private BoundNode RewriteTupleCreationExpression(BoundTupleCreationExpression node, ImmutableArray<BoundExpression> rewrittenArguments)
+        {
+            int remainder;
+            int chainLength = TupleTypeSymbol.NumberOfValueTuples(rewrittenArguments.Length, out remainder);
+
+            NamedTypeSymbol underlyingTupleType = ((TupleTypeSymbol)node.Type).UnderlyingTupleType;
+
+            var currentCreation = default(BoundObjectCreationExpression);
+            int currentArity = remainder;
+            ImmutableArray<BoundExpression> currentArguments = ImmutableArray.Create(rewrittenArguments, (chainLength - 1) * 7, remainder);
+
+            for (int linkIndex = chainLength - 1; linkIndex >= 0; linkIndex--)
+            {
+                if (linkIndex != chainLength - 1)
+                {
+                    currentArity = 8;
+                    currentArguments = ImmutableArray.Create(rewrittenArguments, linkIndex * 7, 7).Add(currentCreation);
+                }
+
+                var ctor = (MethodSymbol)Binder.GetWellKnownTypeMember(_compilation, TupleTypeSymbol.GetTupleCtor(currentArity), _diagnostics, syntax: node.Syntax);
+                if ((object)ctor == null)
+                {
+                    return node;
+                }
+
+                // PERF We are starting with the most nested type, which will involve multiple walking through the underlying tuple type. But long tuples are rarer.
+                NamedTypeSymbol nestedUnderlyingType = TupleTypeSymbol.GetNestedTupleType(underlyingTupleType, linkIndex);
```

Comment:
As we discussed offline, this can be simplified if we push the chain of types into a builder before we enter the loop.